cpp/src/reductions/scan/ewm.cu

```diff
@@ -0,0 +1,374 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "scan.cuh"
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <cuda/functional>
+#include <thrust/scan.h>
+#include <thrust/transform_scan.h>
+
+namespace cudf {
+namespace detail {
+
+template <typename T>
+using pair_type = thrust::pair<T, T>;
+
+/**
+ * @brief functor to be summed over in a prefix sum such that
+ * the recurrence in question is solved. See
+ * G. E. Blelloch. Prefix sums and their applications. Technical Report
+ * CMU-CS-90-190, Nov. 1990. S. 1.4
+ * for details
+ */
+template <typename T>
+class recurrence_functor {
+ public:
+  __device__ pair_type<T> operator()(pair_type<T> ci, pair_type<T> cj)
+  {
+    return {ci.first * cj.first, ci.second * cj.first + cj.second};
+  }
+};
+
+template <typename T>
+class ewma_functor_base {
+ public:
+  T beta;
+  ewma_functor_base(T beta) : beta{beta} {}
+
+  pair_type<T> IDENTITY = {1.0, 0.0};
+};
+
+template <typename T, bool nulls, bool is_numerator>
+class ewma_adjust_functor : public ewma_functor_base<T> {
+  using ewma_functor_base<T>::ewma_functor_base;
+
+ public:
+  using tupletype = std::conditional_t<nulls, thrust::tuple<bool, int, T>, T>;
+
+  __device__ pair_type<T> operator()(tupletype const data)
+  {
+    if constexpr (nulls) {
+      bool const valid = thrust::get<0>(data);
+      int const exp    = thrust::get<1>(data);
+      T const input    = thrust::get<2>(data);
+      T const beta     = this->beta;
+
+      if (!valid) { return this->IDENTITY; }
+
+      T const second = [=]() {
+        if constexpr (is_numerator) {
+          return input;
+        } else {
+          return 1;
+        }
+      }();
+      if (valid and (exp != 0)) {
+        // The value is non-null, but nulls preceding it
+        // must adjust the second element of the pair
+
+        return {beta * (pow(beta, exp)), second};
+      } else {
+        return {beta, second};
+      }
+    } else {
+      if constexpr (is_numerator) {
+        return {this->beta, data};
+      } else {
+        return {this->beta, 1.0};
+      }
+    }
+  }
+};
+
+template <typename T, bool nulls>
+class ewma_noadjust_functor : public ewma_functor_base<T> {
+  using ewma_functor_base<T>::ewma_functor_base;
+
+ public:
+  using tupletype = std::
+    conditional_t<nulls, thrust::tuple<T, size_type, bool, size_type>, thrust::tuple<T, size_type>>;
+
+  __device__ pair_type<T> operator()(tupletype const data)
+  {
+    T const beta          = this->beta;
+    size_type const index = thrust::get<1>(data);
+    T const input         = thrust::get<0>(data);
+
+    if constexpr (!nulls) {
```

Comment:
As above, I think we should use explicit template specializations instead of `if constexpr` here and we can define the right `tupletype` explicitly. Once you do that you can use the structured bindings I suggested above here as well (without that you'd have to copy the index/input unpacking into both specializations as separate lines, but now it's all a one-liner).